### PR TITLE
Update fetch functions and guzzle client instanceissue

### PR DESCRIPTION
**Pull request by Thavarshan**
## **Purpose**

This pull request introduces **v1.1.0** of FetchPHP, addressing key improvements in performance, flexibility, and error handling, while introducing custom middleware support. The most significant change is the deprecation of the `fetchAsync` function in favor of the more consistent `fetch_async`.

**Linked Issues:**
- Improves client efficiency by reusing a singleton Guzzle client.
- Addresses edge cases related to empty or malformed JSON handling.
- Allows developers to add custom middleware for advanced request management (e.g., retries, logging).

---

### **How does this change address the problem?**

- **Singleton Guzzle Client**: By using a static variable, the Guzzle client is instantiated once and reused for multiple requests, reducing overhead. This improves performance, especially in applications making frequent HTTP requests.
- **Custom Middleware Support**: Developers can pass a custom Guzzle client with middleware via the `options['client']` parameter, adding flexibility for logging, retries, and request manipulation.
- **Deprecation of `fetchAsync`**: To improve consistency, `fetchAsync` has been deprecated, and `fetch_async` is introduced. The new naming scheme aligns with best practices and future-proofs the library.
- **Enhanced JSON and Error Handling**: The `Response::json()` method now handles empty or malformed responses more gracefully, throwing exceptions for decoding errors while returning `null` for empty bodies.

---

#### **Open Questions and Pre-Merge TODOs**

- [ ] **Deprecation Handling**: Have we sufficiently warned users about the deprecation of `fetchAsync` in all documentation?
- [ ] **Documentation Updates**: Is the newly added example for custom middleware comprehensive enough?
- [ ] **Thread Safety Considerations**: Should we provide more guidance for using FetchPHP in multi-threaded environments where the static client might introduce concurrency issues?

---

## **Learning**

During the research for this release, we investigated:

1. **Singleton Design Pattern**: Studied the impact of reusing a single Guzzle client to minimize memory usage and reduce the overhead of repeated instantiation.
2. **Custom Middleware Implementation**: Reviewed Guzzle's middleware stack to ensure that FetchPHP could support logging, caching, and retry mechanisms.
3. **Error Handling Best Practices**: Investigated proper handling of JSON decoding errors, ensuring that developers receive clear feedback for empty or malformed responses.

---

This pull request aims to make FetchPHP more robust and versatile while maintaining a smooth migration path for existing users while also pereparing for release of version `1.1.0`. Let us know if any changes or additional feedback are needed before the merge!

### CODE REVIEW

File: tests/FetchTest.php [hunk 1]
```diff
@@ -79,6 +86,23 @@
     expect($response->text())->toBe('Internal Server Error');
 });
 
+// Test for deprecated fetchAsync method (with warning)
+test('fetchAsync shows deprecation warning and can handle a successful GET request', function () {
+    $this->mock->append(new GuzzleResponse(200, ['Content-Type' => 'application/json'], '{"message":"success"}'));
+
+    $promise = fetchAsync('/', [
```
Comment (patinthehat) [inline tests/FetchTest.php:93]: i think this should be `fetch_async`?

File: src/fetch.php [hunk 2]
```diff
@@ -49,77 +70,166 @@ function fetchAsync(string $url, array $options = []): PromiseInterface
...
-            );
+            return handle_async_request($client, $method, $url, $requestOptions);
+        }
+
+        // Handle synchronous request
+        return handle_sync_request($client, $method, $url, $requestOptions);
+    }
+}
+
+if (! function_exists('prepare_body')) {
```
Comment (patinthehat) [inline src/fetch.php:125]: this shouldn't be a global function

**Comment by Thavarshan** [comment]
Closing this in favor of a better refactored PR.

File: src/fetch.php (hunk 2 above)
Comment (Thavarshan) [reply src/fetch.php:125]: Yes, makes sense

File: tests/FetchTest.php (hunk 1 above)
Comment (Thavarshan) [reply tests/FetchTest.php:93]: Well spot!

**Pull request closed without merging**